### ISSUE TITLE
Bug in StringSinkTemplate not reserving enough space in output string.

### ISSUE THREAD

**Issue by Stefan-1313** [weidai11/cryptopp#1064]
I think I stumbled upon a bug in the `StringSinkTemplate`. It seems it is not reserving enough space in the output string.
It tried, but I think the condition is wrong.

I look at `filters.h` line 1145, the if-statement in method Put2: `if (length < size && size + length > m_output->capacity())`.

Program to reproduce:
```
#include <cryptopp/dll.h>
#include <cryptopp/hex.h>
#include <string>


int main(int argc, const char* argv[]) {

	std::cout << "START OF PROGRAM" << std::endl;

	{
		std::string encoded="";
		//encoded.reserve(32);  // ==> If you uncomment this line, the program does not crash any more.
		CryptoPP::byte decoded[] = { 0xFF, 0xEE, 0xDD, 0xCC, 0xBB, 0xAA, 0x99, 0x88, 0x77, 0x66, 0x55, 0x44, 0x33, 0x22, 0x11, 0x00 };
		CryptoPP::HexEncoder encoder2;
		encoder2.Attach(new CryptoPP::StringSink(encoded));
		encoder2.Put(decoded, sizeof(decoded));
		encoder2.MessageEnd();
		std::cout << "Encoded: " << encoded << std::endl;
	}

	std::cout << "END OF PROGRAM" << std::endl;
}

```

So if I don't reserve enough space in the string myself, the program ends up in undefined behavior since it writes outside the reserved memory for the string.

It seems like it is not required for the end-user to do this since it is also never noted in the e.g. [docs of the HexEncoder](https://www.cryptopp.com/wiki/HexEncoder).

This code looks weird to me:
```
typename T::size_type size = m_output->size();
if (length < size && size + length > m_output->capacity())
    m_output->reserve(2*size);
```

I think it should be:
```
typename T::size_type size = m_output->size();
if (size + length > m_output->capacity())  // <= changed this line
    m_output->reserve(size + length);   // <= changed this line
```
First, since the `length` being smaller or greater then the current `size` has nothing to do with the required capacity.
Second, basing the new capacity on the `2*size` seems unlogical, it should be what we check for: `size + length`.

I test this using Visual Studio 2019. And I changed the Runtime Library from `\MTd` to `\MDd` and `\MT` to `\MD` for respectively `Debug` and `Release` build-config. I only tried `x64`. I tested on the released Cryptopp version 8.5.0.

If you agree with the proposed solution I will create a pull request in the up-coming few days.

**Comment by Stefan-1313** [comment]
Maybe the length should be multiplied by 2 since each bite results in two hex characters?
So both in the if-statement as for the reserve-call-argument?

**Comment by Stefan-1313** [comment]
I tested my solution and it did not work. I still got the Exception.

I test this using Visual Studio 2019. And I changed the Runtime Library from `\MTd` to `\MDd` and `\MT` to `\MD` for respectively `Debug` and `Release` build-config. I only tried `x64`.

However, the bug in `StringSink` is still there. If you don't reserve enough space manually in the output string. You will get undefined behavior because `StringSink` writes outside the reserved memory.

Or the doc's of e.g.:

- [HexEncoder](https://www.cryptopp.com/wiki/HexEncoder).
- [Elliptic Curve Integrated Encryption Scheme](https://www.cryptopp.com/wiki/Elliptic_Curve_Integrated_Encryption_Scheme).

Should be updates as these doc's all don't show you have to reserve enough space manually in your string (the doc's just declare an empty string).

**Comment by noloader** [comment]
You have a memory problem. Rerun your tests using the static lib.

Don't use the [FIPS DLL](https://cryptopp.com/wiki/FIPS_DLL). The FIPS DLL is too hard to use correctly. And it is going away soon. Also see [Removal of the FIPS DLL at next Crypto++ release](https://groups.google.com/g/cryptopp-users/c/2GQjimCcACw).

If you want to use a DLL, then create a [Wrapper DLL](https://www.cryptopp.com/wiki/Wrapper_DLL).

**Comment by Stefan-1313** [comment]
I cannot use the static lib as my larger program (in which I want to use Crypto++) is being developed using Dynamic Runtime Linking (`/MD`).
I don't care about 'FIPS compliance' though.

Thanks for your hint tough, I solved the problem. I was using the dll-version mixed with the lib version. This caused the problem somehow. I was messing around now I have it more clearly.

I was put on this astray because of this phrase in the ReadMe.txt which is said to contain all the detailed build instructions:

> To use the Crypto++ DLL in your application, #include "dll.h" before including
> any other Crypto++ header files, and place the DLL in the same directory as
> your .exe file. dll.h includes the line #pragma comment(lib, "cryptopp")
> so you don't have to explicitly list the import library in your project
> settings.

I figured I still needed the DLL because I had to include `dll.h`. And the Visual Studio solution just build it when you `rebuild all`.

Now I build only `Cryptlib.lib` (using Dynamic Runtime Linking (`'\MD`). This works when I don't `#include <cryptopp/dll.h>` (obviously).
